cloud-openstack-server/src/main/java/jetbrains/buildServer/clouds/openstack/OpenstackCloudInstance.java

```diff
@@ -200,15 +202,27 @@ public StartAgentCommand(@NotNull final CloudInstanceUserData data) {
 
         public void run() {
             try {
+                String floatingIp = null;
+                if (cloudImage.isAutoFloatingIp()) {
+                    // Floating ip should be in meta-data before instance start
+                    // If multiple instances start in parallel, perhaps same ip could be retrieved
+                    // So an ip reservation mechanism should implemented in this case
+                    LOG.debug("Retrieve floating ip for future instance association");
+                    floatingIp = cloudImage.getFloatingIpAvailable();
+                    if (StringUtil.isEmpty(floatingIp)) {
+                        throw new OpenstackException("Floating ip could not be found, cancel instance start");
+                    }
+                    LOG.debug(String.format("Floating ip: %s", floatingIp));
+                    userData.addAgentConfigurationParameter(OpenstackCloudParameters.AGENT_CLOUD_IP, floatingIp);
+                }
+
                 String openstackImageId = cloudImage.getOpenstackImageId();
                 String flavorId = cloudImage.getFlavorId();
                 CreateServerOptions options = cloudImage.getImageOptions();
                 options.metadata(userData.getCustomAgentConfigurationParameters());
 
-                // TODO: that code should be in OpenstackCloudImage
-                // but as we make it possible to change userScript
-                // without touching teamcity, that hack takes place
-                // sorry
+                // TODO: that code should be in OpenstackCloudImage but as we make it possible to change userScript without touching teamcity, that
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/axel3rd-github/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=jetbrains.buildServer.clouds%3Acloud-openstack&pullRequest=30&issues=AWTrjQieTfAogIny1gBt&open=AWTrjQieTfAogIny1gBt)

[ISSUE_KEY]: AWTrjQieTfAogIny1gBt